src/index.ts

```diff
@@ -1,183 +1,190 @@
-      | Promise<T>
-      | (() => Promise<T>)
-      | Error
-      | (() => never)
-      | (() => Promise<T>)
-  ): Promise<T> {
-    if (this.hasValue)
-      return new Promise((resolve) => {
-        resolve(this.value);
-      });
-    if (alternative instanceof Error) throw alternative;
-    if (alternative instanceof Function) return alternative();
-    return alternative;
-  }
-
-  valueOrFailure(message?: string): T {
-    if (this.hasValue) return this.value;
-    throw new Error(message ? message : "There exists no value");
-  }
-
-  valueOrUndefined(): T | undefined {
-    if (this.hasValue) return this.value;
-    return undefined;
-  }
-
-  valueOrNull(): T | null {
-    if (this.hasValue) return this.value;
-    return null;
-  }
-
-  toJSON(): T | null {
-    if (this.hasValue) return this.value;
-    return null;
-  }
+    contains(value: T): boolean {
+        if (this.hasValue) return this.value === value;
+        return false;
+    }
+
+    map<TResult>(mapFunc: (value: T) => TResult): Optional<TResult> {
+        return this.match(
+            (value: T) => new Optional<TResult>(mapFunc(value)),
+            () => new Optional<TResult>(undefined)
+        );
+    }
+
+    mapAsync<TResult>(
+        mapFunc: (value: T) => Promise<TResult>
+    ): Promise<Optional<TResult>> {
+        return this.matchAsync(
+            async (value: T) => {
+                var temp = await mapFunc(value);
+                return new Optional<TResult>(temp);
+            },
+            async () => new Optional<TResult>(undefined)
+        );
+    }
+
+    ifNone(func: (() => T) | T): Optional<T> {
+        if (this.hasValue) return new Optional(this.value);
+        else return new Optional<T>(func instanceof Function ? func() : func);
+    }
+
+    async ifNoneAsync(func: (() => Promise<T>) | T): Promise<Optional<T>> {
+        if (this.hasValue) return new Optional(this.value);
+        else return new Optional<T>(func instanceof Function ? await func() : func);
+    }
+
+    filter(predicate: (value: T) => boolean): Optional<T> {
+        if (this.hasValue && predicate(this.value)) return new Optional(this.value);
+        return new Optional<T>(undefined);
+    }
+
+    async filterAsync(
+        predicate: (value: T) => Promise<boolean>
+    ): Promise<Optional<T>> {
+        if (this.hasValue && (await predicate(this.value)))
+            return new Optional(this.value);
+        return new Optional<T>(undefined);
+    }
+
+    flatMap<TResult>(
+        flatMapFunc: (value: T) => Optional<TResult>
+    ): Optional<TResult> {
+        return this.match(flatMapFunc, () => new Optional<TResult>(undefined));
+    }
+
+    flatMapAsync<TResult>(
+        flatMapFunc: (value: T) => Promise<Optional<TResult>>
+    ): Promise<Optional<TResult>> {
+        return this.match(
+            flatMapFunc,
+            async () => new Optional<TResult>(undefined)
+        );
+    }
+
+    ifPresent(consumer: (value: T) => void) {
```

Comment:
Consumer should be of type: 
```suggestion
    ifPresent(consumer: (value: A) => B) {
```

Then B can both be void or something else, ensuring we're not limiting or care about the function that gets passed to us.